Content.Server/Bar/PolymorphGlassSystem.cs

```diff
@@ -0,0 +1,103 @@
+using Robust.Shared.Prototypes;
+using Robust.Server.GameObjects;
+using Content.Shared.Containers.ItemSlots;
+using Content.Shared.FixedPoint;
+using Content.Shared.Bar;
+
+namespace Content.Server.Bar;
+
+public sealed class TransformableContainerSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly SharedSolutionContainerSystem _solutionsSystem = default!;
+    [Dependency] private readonly ItemSlotsSystem _itemSlotsSystem = default!;
+    [Dependency] private readonly TransformSystem _xformSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<PolymorphGlassComponent, GetVerbsEvent<Verb>>(OnGetVerb);
+    }
+
+    private void ChangeGlass(EntityUid uid, PolymorphGlassComponent component, EntityPrototype prototype)
+    {
+        if (!_solutionsSystem.TryGetSolution(uid, "drink", out var sourceSolutionEntity, out var sourceSolution))
+            return;
+
+        var position = Transform(uid).Coordinates;
+        var targetEntity = Spawn(prototype.ID, position);
+
+        if (!_solutionsSystem.TryGetSolution(targetEntity, "drink", out var targetSolutionEntity, out var targetSolution))
+        {
+            Del(uid);
+            return;
+        }
+
+        _solutionsSystem.TryTransferSolution(targetSolutionEntity.Value, sourceSolution, 30.0f);
+
+        if (!TryComp<ItemSlotsComponent>(uid, out var sourceItemSlots))
+        {
+            Del(uid);
+            return;
+        }
+
+        if (!_itemSlotsSystem.TryGetSlot(uid, "garnish_slot", out var sourceGarnishSlot) ||
+            !_itemSlotsSystem.TryGetSlot(uid, "decoration_slot", out var sourceDecorationSlot))
+        {
+            Del(uid);
+            return;
+        }
+
+        var outGarnish = _itemSlotsSystem.TryEject(uid, sourceGarnishSlot, null, out var sourceGarnish, true);
+        var outDecoration = _itemSlotsSystem.TryEject(uid, sourceDecorationSlot, null, out var sourceDecoration, true);
+
+        if (TryComp<ItemSlotsComponent>(targetEntity, out var targetItemSlots))
+        {
+            if (!_itemSlotsSystem.TryGetSlot(targetEntity, "garnish_slot", out var targetGarnishSlot) ||
+                !_itemSlotsSystem.TryGetSlot(targetEntity, "decoration_slot", out var targetDecorationSlot))
+            {
+                Del(uid);
+                return;
+            }
+
+            if (outGarnish && sourceGarnish.HasValue)
+                _itemSlotsSystem.TryInsert(targetEntity, "garnish_slot", sourceGarnish.Value, null, targetItemSlots, true);
+            if (outDecoration && sourceDecoration.HasValue)
+                _itemSlotsSystem.TryInsert(targetEntity, "decoration_slot", sourceDecoration.Value, null, targetItemSlots, true);
+        }
+
+        Del(uid);
+    }
+
+    private void OnGetVerb(EntityUid uid, PolymorphGlassComponent component, GetVerbsEvent<Verb> args)
```

Comment:
Is it even worth it though? If I put the system in shared it bugs out the game (it creates client side only entities). A non-predicted verb of a glass doesn't seem like it will bother anyone.

Reply:
Yeah your doing it a little wrong then, it shouldn't be calling any function that creates entities in client! You have to make an overridable function where if you run it in client, it does nothing, but in server it actually does the stuff.

I personally think its important because having verbs pop up really messes when you click sometimes. However, probably only bother doing this at the end because it isn't like vital for the PR (Figuring out how the glasses are going to work gong to be much more difficult.)

Reply:
Unless a maintainer tells me it's necessary I don't think i'll go through the work.

Reply:
Nevermind, i guess i did it?

Reply:
Haven't tested it but that's what I was saying! Looks good 🫡